core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -70,18 +73,34 @@
     private static final int BUCKET_SIZE = 8;
     private static final int MASK = BUCKET_SIZE  - 1;
 
+    @Nullable
+    private final AttributeMap root;
+
     // Initialize lazily to reduce memory consumption; updated by AtomicReferenceFieldUpdater above.
     @VisibleForTesting
+    @Nullable
     volatile AtomicReferenceArray<DefaultAttribute<?>> attributes;
 
+    /**
+     * Creates a new instance.
+     */
+    public DefaultAttributeMap(@Nullable AttributeMap root) {
+        this.root = root;
+    }
+
     @SuppressWarnings("unchecked")
     @Override
     public <T> Attribute<T> attr(AttributeKey<T> key) {
         if (key == null) {
             throw new NullPointerException("key");
         }
+
         AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
         if (attributes == null) {
+            if (root != null && root.hasAttr(key)) {
+                return root.attr(key);
```

Comment:
How about copying the attribute here, so that any attribute changes do not affect the root? e.g.

```java
DefaultAttributeMap root = new DefaultAttributeMap(null);
root.attr(FOO).set("bar");
DefaultAttributeMap map = new DefaultAttributeMap(map);
map.attr(FOO).set("baz");
assert root.attr(FOO).get()).equals("bar");
assert map.attr(FOO).get()).equals("baz");
```

Alternatively, we could return an `Attribute` wrapper whose `set()` will update `map` instead of `root`, but maybe that's more complex?